src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudClient.java

```diff
@@ -0,0 +1,234 @@
+   * @throws ETSdkException The FuelSDKException
+   */
+  public static MarketingCloudClient create(String clientId, String clientSecret, String authEndpoint,
+                                            String soapEndpoint) throws ETSdkException {
+    ETConfiguration conf = new ETConfiguration();
+    conf.set("clientId", clientId);
+    conf.set("clientSecret", clientSecret);
+    conf.set("authEndpoint", authEndpoint);
+    conf.set("soapEndpoint", soapEndpoint);
+    conf.set("useOAuth2Authentication", "true");
+    ClassLoader oldCL = Thread.currentThread().getContextClassLoader();
+    try {
+      Thread.currentThread().setContextClassLoader(MarketingCloudClient.class.getClassLoader());
+      return new MarketingCloudClient(new ETClient(conf));
+    } finally {
+      Thread.currentThread().setContextClassLoader(oldCL);
+    }
+  }
+
+  /**
+   * Check if filter string is a valid SFMC Expression.
+   *
+   * @param filter
+   * @return
+   */
+  public static void validateFilter(String filter) throws ETSdkException {
+    if (!Util.isNullOrEmpty(filter)) {
+      ETExpression expression = ETExpression.parse(filter);
+    }
+  }
+
+  private ETExpression getExpressionfromString(String filter) throws ETSdkException {
+    if (!Util.isNullOrEmpty(filter)) {
+      return ETExpression.parse(filter);
+    } else {
+      return new ETExpression();
+    }
+  }
+
+  /**
+   * Fetch records for passed object from Salesforce Marketing Cloud.
+   *
+   * @param object The SourceObject which tells what data to be fetched from Salesforce Marketing Cloud
+   * @return The list of ETApiObject representing the records from requested object
+   */
+  public List<? extends ETSoapObject> fetchObjectRecords(SourceObject object) throws ETSdkException {
+    ETFilter filter = new ETFilter();
+    filter.setExpression(getExpressionfromString(object.getFilter()));
+    return fetchObjectData(client, object.getClassRef(), filter);
+  }
+
+  /**
+   * Fetch the schema information for passed object from Salesforce Marketing Cloud.
+   *
+   * @param object The SourceObject which tells for which schema to be fetched from Salesforce Marketing Cloud
+   * @return The instance of MarketingCloudObjectInfo object
+   * @throws ETSdkException The FuelSDKException
+   */
+  public MarketingCloudObjectInfo fetchObjectSchema(SourceObject object)
+    throws ETSdkException {
+    Class<? extends ETSoapObject> clazz = object.getClassRef();
+    return new MarketingCloudObjectInfo(object, fetchObjectFields(clazz));
+  }
+
+  /**
+   * Fetch records for passed object from Salesforce Marketing Cloud.
+   *
+   * @param dataExtensionKey The data extension key for which data to be fetched from Salesforce Marketing Cloud
+   * @return The list of ETDataExtensionRow representing the records from requested data extension
+   * @throws ETSdkException The FuelSDKException
+   */
+  public List<ETDataExtensionRow> fetchDataExtensionRecords(String dataExtensionKey, String filterStr)
```

Comment:
looks like this loads the entire contents of the data extension into memory. I would expect this to go out of memory for any non-tiny data extension. What size data extensions was this tested on?

Reply:
On a 1CPU/ 2GB worker,  client successfully retrieved about 1.6M rows from Data extension.